Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -178,35 +202,50 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
         return;
     }
 
-    PAL::CMBufferQueueEnqueue(m_compressedSampleQueue.get(), cmSampleBuffer);
-    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }] {
-        if (RefPtr protectedThis = weakThis.get())
+    ++m_framesBeingDecoded;
+    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, sample = RetainPtr { cmSampleBuffer }, flushId = m_flushId.load()]() mutable {
+        if (RefPtr protectedThis = weakThis.get()) {
+            assertIsCurrent(protectedThis->m_workQueue.get());
+
+            if (flushId != protectedThis->m_flushId) {
+                protectedThis->m_compressedSampleQueue.clear();
+                return;
+            }
+
+            protectedThis->m_compressedSampleQueue.append({ WTFMove(sample), flushId });
             protectedThis->decodeNextSample();
+        }
     });
 }
 
 void VideoMediaSampleRenderer::decodeNextSample()
 {
-    assertIsCurrent(m_workQueue);
-
-    ASSERT(m_compressedSampleQueue);
+    assertIsCurrent(m_workQueue.get());
 
     if (m_isDecodingSample || m_gotDecodingError)
         return;
 
-    auto sample = adoptCF(checked_cf_cast<CMSampleBufferRef>(PAL::CMBufferQueueDequeueAndRetain(m_compressedSampleQueue.get())));
-    if (!sample)
+    if (m_compressedSampleQueue.isEmpty())
         return;
 
+    auto [sample, flushId] = m_compressedSampleQueue.takeFirst();
+    if (flushId != m_flushId)
+        return decodeNextSample();
+
+    --m_framesBeingDecoded;
+    ASSERT(m_framesBeingDecoded >= 0);
+
     bool displaying = !MediaSampleAVFObjC::isCMSampleBufferNonDisplaying(sample.get());
     auto decodePromise = m_decompressionSession->decodeSample(sample.get(), displaying);
     m_isDecodingSample = true;
-    decodePromise->whenSettled(m_workQueue, [weakThis = ThreadSafeWeakPtr { *this }, displaying](auto&& result) {
+    decodePromise->whenSettled(m_workQueue, [weakThis = ThreadSafeWeakPtr { *this }, displaying, flushId = flushId](auto&& result) {
         RefPtr protectedThis = weakThis.get();
         if (!protectedThis)
             return;
 
         protectedThis->m_isDecodingSample = false;
+        if (flushId != protectedThis->m_flushId)
+            return;
```

Comment:
It would be, but I hinted at this in the commit log. Making WebCoreDecompressionSession return in PTS order frame, we need to increase its latency, and we wouldn't be able to keep the simple 1:1 ratio between call to decode and having a one frame out